### PR TITLE
interfaces:ensure that implementations are ordered

### LINKED ISSUE #11762: Evaluate template & interface preconditions one by one
Currently we evaluate all preconditions first and then check the result. I think it makes more sense to switch this to evaluate the preconditions one by one and fail on the first one that doesn’t hold.

### PR DESCRIPTION
**Pull request by robin-da**
We make sure that the implementations are ordered by insertion in the
Scala AST. This is important to guarantee an evaluation order of the
interface preconditions that is determined by the order of interface
implementations of a template.

Fixes #11762.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
As discussed, we could also enforce the ordering of implementations more explicitly in the data structures of our ghc fork.

**Review by cocreature — APPROVED** [review]
Looks great thank you! @remyhaemmerle-da can you please also take a look

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala [hunk 1]
```diff
@@ -745,7 +746,9 @@ object Ast {
       choices: Map[ChoiceName, GenTemplateChoice[E]], // Choices available in the template.
       observers: E, // Observers of the contract.
       key: Option[GenTemplateKey[E]],
-      implements: Map[TypeConName, GenTemplateImplements[E]],
+      implements: ListMap[TypeConName, GenTemplateImplements[
```
Comment (remyhaemmerle-da) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala:749]: Unfortunately `ListMap` has bad complexity. See [Scala Std Library Performance Characteristics](https://docs.scala-lang.org/overviews/collections/performance-characteristics.html). 
I think we have to store the order as a separate field, or revive `InsertOrderMap` .

**Review by remyhaemmerle-da — COMMENTED** [review]
I am worried we degrade the  complexity of `implements` lookup, and indirectly choice lookup.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala:749]: I think I’d go for restoring InsertOrderMap. Seems easier to maintain than to keep the order in a separate field.

Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala:749]: Let's do that in another PR.

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala (hunk 1 above)
Comment (robin-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala]: I replaced ListMap with VectorMap, which should give us amortized O(1) performance on the lookups and preserves insertion order. An alternative that also preserves insertion order is TreeSeqMap.

Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala]: Good finding. I was not aware of this addition to scala 2.13. 
Thanks.

**Pull request merged**